### PR TITLE
Blog Post: The Worst Decision Of My Career

### PR DESCRIPTION
**Pull request by gabrielpoca**

### CODE REVIEW

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+Slow down! Accountants invented systems around manual work. Those systems were designed to prevent human error. We still have to account for human error in software, but it's a different kind of error, so we need a different kind of solution.
+
+Event-sourcing indeed solves a lot of problems, but there are other, simpler, solutions that you can use. You can keep track of the changes in a table (on your database) with just another table and some triggers. If you're into Rails, [paper_trail](https://github.com/paper-trail-gem/paper_trail) is the tool for you.
+
+Is your data time-series? No problem. Put those events to a table and use a materialized view to read the data. Small steps can take you a long way.
+
+## Banking
```
Comment (frm) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:119]: Are you sure this and `Making the right call` have the correct heading level?

Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:119]: True! Somehow I messed this up again.

**Comment by gabrielpoca** [comment]
@pfac I'm honestly out of ideas :/ I already spent so many hours in this, if you think it's fine, let's merge it as it is

**Comment by pfac** [comment]
> @pfac I'm honestly out of ideas :/ I already spent so many hours in this, if you think it's fine, let's merge it as it is

Sure thing, feel free to merge it when you're ready.

**Comment by zamith** [comment]
Let's start by saying that I don't completely agree with the main point of the article, in the particular case of utrust (which is never mentioned, but is abundantly clear for people in the know), I think that event sourcing for the payment flow was one of the best decisions we made. Of course, as with any new technology, we overdid it at first, using it in places where we shouldn't have (such as user management) and had to learn from that.

In terms of the article itself, I feel it's very opinionated, but that's not necessarily bad, I've written opinionated articles for this blog before. The main issue I have with it is that it is opinionated in a negative way. It not only gives your opinion, but it goes at length (probably too much) to try and "destroy" another point of view (that is still shared by a lot of people in the Sub universe). 

In sum, I think it's ok to write an article about your experience with these and even that early on you should shy away from complexity. However, I don't think we should publish an article that (to me, at least) comes across as so negative not to only to yourself (or us), but also other people we know and in general. I think you can make your point in a more positive way.

**Comment by gabrielpoca** [comment]
@zamith I understand your point. I made a couple of changes to try and address what you mentioned. Would you mind looking at it again? I know the article will come across as negative, but that's how I want it to be. I know a lot of people in the universe will disagree, but a lot more will see themselves in my words and my pain. I'm ok with a little sanitization to my ideas, but to be honest, I would rather publish this in my personal blog than not being able to convey this message.

**Comment by zamith** [comment]
Not sure it's my call to make, but I don't think this aligns with the tone in our blog, nor with the message we want to put across as part of our OKRs.

**Comment by gabrielpoca** [comment]
@zamith not even with my latest changes?

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+confident. As if there was nothing we couldn't do. We've already learned a lot,
+so we know what to avoid. Never again, we said!
+
+So what was the worst decision of my career? It's clear to me - although some
+will disagree - that the worst decision I made was starting a product with
+event-sourcing, asynchronous communications, event-driven architecture, and all
+of that. Mind you, I didn't make the decision alone, but that doesn't change
+anything.
+
+Why was that the worst decision? Because we didn't know the market. We didn't
```
Comment (zamith) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:123]: I kind of agree with the idea that because of it being harder to change than a "regular" system, event sourced systems might not be a great solution when you are still learning the field. You want to optimize for rate of change and ES prevents it somewhat. I really don't like calling it the "worst decision ever". I think it's too much of a hyperbole. 

What about the decision of letting Oncostats linger for years with no end in sight? Or other things we built, knowing perfectly well they were a waste of money? Those are also decision you (and all of us) made not to stand your ground and stop them on their tracks. I guess you mean technical decision (even though you never say it), but even then, what about our previous website for example, when we were experimenting with SUITCSS and went completely crazy? 

I get that you have strong feelings about this one, but I don't think it is the worst decision of your career, nor do I think it should be portrayed that way. It's a click baity title that is more likely to offend people we know than to attract people we don't.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+[paper_trail](https://github.com/paper-trail-gem/paper_trail) is the tool for
+you.
+
+Is your data time-series? No problem. Put those events to a table and use
+a materialized view to read the data. Small steps can take you a long way.
+
+### Banking
+
+If you happen to have event-sourcing enthusiasts in your team, they will jump
+to the opportunity to implement a payment system because banks use
```
Comment (zamith) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:245]: I think the argument has always been that conceptually what banks do is like ES, but mostly implemented in an ad hoc way that's not as thought out.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+Am I going to use event-sourcing or microservices again? Sure, it depends on
+the context and the client. The decisions I make are the best that I can with
+the information that I have. For instance, when I'm part of the team that's
+starting a product, the technology we pick will depend on how we'll hire: if
+you want to build an office in Braga, we have to make sure we have developers
+for that technology available. We have to think things through, and some things
+you only learn from experience. This applies to the systems we design as well.
+I've seen enough people design around what they believe the product will become
+in two years to know that those designs always fail to accommodate the changes
+that will come. So I design systems for small, incremental changes. Do the
```
Comment (zamith) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:269]: If this is to be published in the Subvisual blog I think you can/should change this to "we". These are principles that I believe everyone shares and also a good lead to a "Hire us" call to action.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+Most of us will get bored without novelty. That's why the software industry has
+a recycling mechanism to keep things fresh and interesting for us: every once
+in a while a new, or different, language/framework will rise and light the path
+for a brighter future, overthrowing the existing standard, demanding that we
+learn it to be on top of the game.
+
+We change our tools, but we keep building the same things over and over. I've
+been doing this long enough to see that we are just going in circles. Let's be
+honest, most of us aren't building life-changing products that wouldn't have
+been possible ten years ago, so let's not jump to the sight of another shiny
```
Comment (zamith) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:47]: I don't think "jump to the sight" is an expression. Maybe "jump straight into another"?

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+> If all you have is a hammer, everything looks like a nail _Abraham Maslow_
+
+The designers of Erlang picked the right tool for the job, and most of us want
+to do the same. Unfortunately, there's usually more than one "tool" that would
+be "right," but to know which ones, you need to know what "job" you're solving.
+All of us **should** know this, but I keep learning about teams that fail to do
+it. There are so many companies, with a handful of "experienced" developers,
+that don't know what they are building, don't have a clear business yet, but
+their product is already built on complex architectures and technologies such
+as microservices, Kubernetes or event-sourcing. I know why we make these
```
Comment (zamith) [inline src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: If were bashing on things, I would add SPAs here :) Or anything that means to deal with offline as well.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+I was going to write a whole section on what makes for a good starting team,
+but that's a different topic, and there's still so much I need to say.
+
+## The decision
+
+I've made a lot of bad decisions, and I've been fortunate enough to suffer the
+consequences of those decisions. A lot of developers don't get to experience
+that, so they never learn. That is also why I'm always skeptical about external
+"expert" consultants; they usually don't stick around to experience the
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:110]: I agree 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+of duplication from our projections because they were unmaintainable.
+
+With time, everything became more manageable. We defined clear APIs for our
+services (umbrella apps) that allowed us to extract three other apps later
+(Elixir is great for this kind of thing). We wrote integration tests that would
+run with confidence. We even created custom linters to catch changes that would
+typically lead to errors. Everything was falling into place.
+
+But the mess that we made still showed up in many places. For instance, our
+APIs were eventually consistent, which meant that our front-end had to know
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:152]: I agree that eventually consistent APIs aren't always an issue, but the example I provide explains when I think it's an issue. I can rephrase if the message isn't clear, but I don't think we can separate the eventually consistent behaviors from the APIs, they live together. Even if the payment API was, in theory, an example of when eventually consistency is fine, we had to built a loading screen on the widget in order to wait for the data that was just submitted by the merchant to create a payment.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+that when it makes a request to create a resource, and the response is
+successful, it doesn't mean that the resource was created, just the request to
+create was successful. If for some reason, the front-end needs to make another
+request to ask for that resource, it could fail because it didn't reach the
+projections yet. Because of this, we had to make a mess on the codebase of the
+front-end as well. This is what happens when you take the responsibility of
+consistency in your own hands. And don't get me started in side-effects in the
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:159]: I understand the argument that the problem was due to our inexperience, but I don't agree that inexperience is the main issue. Sure, it helps when you already know what problems you'll encounter, but it doesn't change the fact that there's no good solution for them. The point here is that we bought these issues.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 6 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: I don't think SPAs or offline are at the same level and I would rather not diverge from the main discussion.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+chain of events.
+
+Event-sourced systems are complex, and that's fine; as I said, we are in the
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:162]: I'm not sure what your point is on this one. SPAs with Redux are on the same level of complexity when you have offline storage that you cannot clear when the user refreshes the page. I feel that say that it's complex as a follow up to what I've said before is safe enough.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+Some people have a way with words that can make anything sound like the best
+idea ever. If you add some authority to the mix, no one will stand a chance
+until they make a mistake, or are proven wrong with facts. Not that you need to
+push yourself too much to make event-sourcing and microservices sound great; in
+theory, it's the best thing ever!
+
+I consider myself a professional skeptic, and even that didn't save me from
+falling into this trap. But maybe I can help you, so in the next sections, I'll
+analyze some of the arguments that I've heard to support event-sourcing and
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:179]: But I'm not trying to say they go hand in hand. I'm trying to treat them as separate. Maybe I can rephrase this to make it clearer.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 3 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:245]: But in that case, in pratice banks could be using an events table with a materialized view?

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 4 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:269]: 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 2 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:123]: I actually like the title a lot :p but sure, there are strong contenders to the worst decision title, but it doesn't seem like a stretch for it to be this one.

What's the matter with SUITCSS? It's still our standard for non-component systems. We used it in many places and even now we are teaching it on the summer camp.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 5 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:47]: 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+_This is awesome, so let's do the same thing in code now! Besides, if you never
+delete anything, you can always go back and look, and maybe use that data for
+other purposes._
+
+Slow down! Accountants invented systems around manual work. Those systems were
+designed to prevent human error. We still have to account for human error in
+software, but it's a different kind of error, so we need a different kind of
+solution.
+
+Event-sourcing indeed solves a lot of problems, but there are other, simpler,
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:233]: I understand that, but it doesn't change the argument. Every business event that you record with events can be recorded in other ways. I mention time-series in the next paragraph, but I'll add to it to give that example as well 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+know any better, the event store, which kept the state of **all** services, was
+also being used as an event-bus to communicate between those services. Because
+of that, it was possible to couple one service to the internal state of
+another, which happened, all over the place, all the time. This almost
+invisible coupling made everything worse.
+
+## Aftermath
+
+As with any new technology (for us), we overdid it at first, and because of
+that we were lost. Every change increased the number of bugs. We spent long
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:139]: "over our heads" maybe?

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:162]: What I meant was the I think complexity should always be posed as relative to something else. For example, "Event-sourced systems are, in general, more complex than a CRUD system"

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 2 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:123]: We went completely overboard with it, to a point that was crazy (in my opinion). If you remember we needed like 1 thousands divs to write a title, because of placing things and stuff like that.

I'm not a fan of the title a all, to be honest. It's very negative and I think the Subvisual's blog message should be a positive one, even when talking about when we mess up.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:159]: I think it's similar to the "JS the good parts" book. Because it has many pitfalls, does that mean using it in the first place is an error?

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 3 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:245]: Sure, they can. I have no idea what they're doing.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 6 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: I think they totally are, tbh. I think that you are way more proficient with them, so you don't see it.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:139]: I think that also implies that we were going around not really knowing what to do. I both don't think it depicts a true image of what happened and is not something nice for us.

[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:139]: But it does, I remember clearly a time when everyone was in Braga, Maciek, Zdenko, Antonin, and we were all talking about this. We were lost. Integration tests were failing randomly and we didn't know why. Commits were breaking the build because of data migrations. The front-end was reporting weird errors. It is a true image. But I understand that we may see things differently. The message I want to convey is that too many things went wrong that we didn't expect. Do you have a more appropriate suggestion?

[the remark this answers is not in the document]
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:139]: That phrasing is better. Something along the lines of "As with any new tech... there were things that went wrong that we didn't antecipate."

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:159]: Right, I get that. And I think that for someone that's been through it with utrust can relate, but I'm not sure the takeaways on why it was a bad decision (and I maintain that for the payments it was net good) are conveyed to someone on the outside, which I assume is going to be most of the people reading this.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 2 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:123]: Yes, it's only my opinion, not sure what other people think about it.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 6 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: Not what I was trying to do. What I meant was that when you are comfortable with a technology, it seems simpler to you than it probably is to beginners. It happens a lot with the "Rails magic" thing, I don't see it as magic because I'm so familiar with how things work that it makes total sense to me.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:159]: ok, I think I can work with that feedback 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 6 above)
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: I don't want to add SPAs and offline to this list because it makes no sense to me. I'm usually more vocal regarding front-end stuff, but it isn't even what I have more experience with: I spent the last two years working in Elixir and dealing with ES issues, DevOps and others; and before that, I was working with Ruby on Nourish's events system, where I used materialized views and triggers and all of that. Sure, I'm not that experienced in microservices and Kubernetes, but I'm not a beginner because of the experience I have on the technologies around those. We'll just go in circles, I know you have a special hate for JS and SPAs and all of that, but I really don't want to have anything to do with it in this blog post.

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:139]: 👍

File: src/posts/2020-07-30-the-worst-decision-of-my-career.md (hunk 6 above)
Comment (zamith) [reply src/posts/2020-07-30-the-worst-decision-of-my-career.md:81]: Well, I think the opposite. They are all examples of overly complex things people tend to reach for wait too early in projects, in my opinion. But let's agree to disagree, then.

**Comment by gabrielpoca** [comment]
Closing for now. Will rewrite some sections soon.
